src/wasm/wasm-validator.cpp

```diff
@@ -276,11 +277,52 @@ struct FunctionValidator : public WalkerPass<PostWalker<FunctionValidator>> {
   void visitLoop(Loop* curr);
   void visitIf(If* curr);
 
+  static void visitPreTry(FunctionValidator* self, Expression** currp) {
+    auto* curr = (*currp)->cast<Try>();
+    if (curr->name.is()) {
+      self->delegateTargetNames.insert(curr->name);
+    }
+  }
+
+  // We remove try's label before proceeding to verify catch bodies because the
+  // following is a validation failure:
+  // (try $l0
+  //   (do ... )
+  //   (catch $e
+  //     (try
+  //       (do ...)
+  //       (delegate $l0) ;; validation failure
+  //     )
+  //   )
+  // )
+  // Unlike branches, if delegate's target 'catch' is located above the
+  // delegate, it is a validation failure.
```

Comment:
To be sure I understand, is this because it could be an infinite loop otherwise?

Reply:
That can be one reason, but I haven't thought about that actually. The motivation (for the spec) was just it was unnatural both intuitively and for the VM implementation to delegate upwards; it looks like we are branching upwards.